src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -844,6 +826,41 @@ private EnumeratorResult GetEnumeratorInfo(ref ForEachEnumeratorInfo.Builder bui
             return EnumeratorResult.FailedNotReported;
         }
 
+        private void GetDisposalInfoForEnumerator(ref ForEachEnumeratorInfo.Builder builder, BoundExpression expr, bool isAsync, DiagnosticBag diagnostics)
+        {
+            // NOTE: if IDisposable is not available at all, no diagnostics will be reported - we will just assume that
+            // the enumerator is not disposable.  If it has IDisposable in its interface list, there will be a diagnostic there.
+            // If IDisposable is available but its Dispose method is not, then diagnostics will be reported only if the enumerator
+            // is potentially disposable.
+
+            TypeSymbol enumeratorType = builder.GetEnumeratorMethod.ReturnType.TypeSymbol;
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+
+            if (!enumeratorType.IsSealed ||
+                this.Conversions.ClassifyImplicitConversionFromType(enumeratorType,
+                    isAsync ? this.Compilation.GetWellKnownType(WellKnownType.System_IAsyncDisposable) : this.Compilation.GetSpecialType(SpecialType.System_IDisposable),
+                    ref useSiteDiagnostics).IsImplicit)
+            {
+                builder.NeedsDisposeMethod = true;
+            }
+            else if (enumeratorType.IsValueType && enumeratorType.IsRefLikeType)
+            {
+                // if it wasn't directly convertable to IDisposable, see if it is pattern disposable
+                // again, we throw away any binding diagnostics, and assume it's not disposable if we encounter errors
+                DiagnosticBag patternDisposeDiags = new DiagnosticBag();
```

Comment:
nit: consider using `var` here and next line, but explicit type on following line (`disposeMethod`)